Azure/sap-hana
# Application input configuration for VM size and accelerated networking and OS helper script update

# Problem

As logged in #517 the application Tier VMs should have the Size determined by a variable that the user can override.

Additionally the `util/set_sap_os.sh` helper script does not operate on the application configuration.

## Description

The size for the Application and SCS VMs is now determined via variable which the user can override in the input configuration.  As not all VM SKUs support Accelerated networking, this option is tied to the VM configuration.

[The current default for both App and SCS VMs is `Standard_D4s_v3`, and accelerated networking is turned off.](https://github.com/Azure/sap-hana/pull/551/files#diff-6cdfef3730721bb9a457716a667c0c77R39-R45)

This can be overridden by providing an object for `application.vm_config` in the input JSON file.  The object takes two keys, `scs` and `app`.  Each is a comma separated string of the VM SKU, and `true` or `false` for the accelerated networking parameter.  This is documented in the [JSON Adjusting Template document](https://github.com/Azure/sap-hana/pull/551/files#diff-d346278c57da72080e62e06c2648c2d9R147-R148).

The `util/set_sap_os.sh` script has been updated with an optional third parameter of `app` or `hdb`. This third parameter will limit the OS Image change to the Application or HANA Database input configuration.  If no third param is provided to the script (i.e. existing usage) both sections are updated.

# Test Instructions/Acceptance Criteria

1. Review the code changes:
   - [ ] [Input JSON changes are acceptable](https://github.com/Azure/sap-hana/pull/551/files#diff-87e07c611a9c32b818e759bdb4b172ffR203-R206)
   - [ ] [JSON Template documentation sufficiently updated](https://github.com/Azure/sap-hana/pull/551/files#diff-d346278c57da72080e62e06c2648c2d9R147-R148)
   - [ ] [Defaults are sensible and can be overridden by the user](https://github.com/Azure/sap-hana/pull/551/files#diff-6cdfef3730721bb9a457716a667c0c77R38-R45)
   - [ ] [App and SCS configuration is clearly sourced](https://github.com/Azure/sap-hana/pull/551/files#diff-6cdfef3730721bb9a457716a667c0c77R47-R51)
   - [ ] [App NIC and VM use new configuration](https://github.com/Azure/sap-hana/pull/551/files#diff-1e733d3d2860b994aa9e9761363a5e69R7-R41)
   - [ ] SCS NIC and VM use new configuration](https://github.com/Azure/sap-hana/pull/551/files#diff-e2d161cdc955d1fa4ec5b42040f1ee6cR7-R115)
1. Test the SAP OS script:
   On a new clean checkout of the code:
   1. `util/set_sap_os.sh`
      - [ ] Usage is output to screen, describes third parameter
   1. Update the Single Node HANA template to RHEL:\
      `util/set_sap_os.sh rhel single_node_hana`
   1. Check the diff:\
      `git diff deploy/template_samples/single_node_hana.json`
      - [ ] The OS sections for both `databases` and `application` are updated to RHEL
   1. Update the SNH template to SLES for HDB:\
      `util/set_sap_os.sh sles single_node_hana hdb`
   1. Check the diff:\
      `git diff deploy/template_samples/single_node_hana.json`
      - [ ] The OS section for `databases` is no longer listed as it has been reverted to SLES
   1. Update the SNH template to SLES for the application:\
      `util/set_sap_os.sh sles single_node_hana app`
   1. Check the diff:\
      `git diff deploy/template_samples/single_node_hana.json`
      - [ ] The OS section for `application` is no longer listed as it has been reverted to SLES
